### PR TITLE
Callstack visualization

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/1472.

### CODE REVIEW

**Comment by SouthEndMusic** [comment]
Obtaining the call graph seems to work well. We should just restrict it to smaller parts of the code, not an entire run. For visualization I went with a `Makie` approach, here is an example for `Parameters`:

image

We can debate very long about what the best way would be to visualize the data, but because it gets very dense very quickly I think this plot is best used interactively. I do think the color coding by script is quite nice.

File: docs/dev/callstacks.quarto_ipynb
Comment (evetion) [inline docs/dev/callstacks.quarto_ipynb:1]: Can be removed as it is generated by quarto

File: Project.toml [hunk 1]
```diff
@@ -20,6 +21,7 @@ DiffEqCallbacks = "459566f4-90b8-5000-8ac3-15dfb0a30def"
 EnumX = "4e289a0a-7415-4d19-859d-a7e5c4648b56"
 FindFirstFunctions = "64ca27bc-2ba2-4a57-88aa-44e436879224"
 FiniteDiff = "6a86dc24-6348-571c-b903-95158fe2bd41"
+GLMakie = "e9467ef8-e4e7-5192-8a1a-b1aee30e663a"
```
Comment (evetion) [inline Project.toml:24]: This is useful for interaction, but I can imagine CairoMakie is enough for quarto static plots?

**Review by evetion — COMMENTED** [review]
Great effort! 💯 

Would it be useful to depend on https://graph.makie.org/stable/generated/depgraph/? The current layout seems quite constrained, ideally you have an automatic graph layout.

For interactivity, a script in utils might be a better fit? Or a new util folder in docs, so the quarto notebook is a bit shorter?

Did you test this with a single calculation timestep as well?

Finally, given your slack question, any progress on an mermaidjs translation?

**Comment by SouthEndMusic** [comment]
Improved formatting, generated functions cut out:

image

**Comment by SouthEndMusic** [comment]
The graph for a run of the allocation algorithm:

image

I also want to show allocation initialization but for some reason the call trace code crashes on that at the moment.

**Comment by SouthEndMusic** [comment]
Ah I forgot `DiscreteControl`

File: docs/dev/callstacks.qmd [hunk 2]
```diff
@@ -0,0 +1,413 @@
+# Call stacks
```
Comment (evetion) [inline docs/dev/callstacks.qmd:1]: Maybe move all this code into a separate .jl file and `include` that here. That makes re-use easier, and also paves the way for creating CallGraph.jl in the future.

File: docs/dev/callstacks.qmd [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+    squash_methods = [
+        :n_neighbor_bounds_flow,
+        :n_neighbor_bounds_control,
+        :sort_by_function,
+        :neighbortypes
+    ],
+    xlims = (-0.4, 5.6)
+)
+```
+
```
Comment (evetion) [inline docs/dev/callstacks.qmd:361]: Maybe add the update_cumulative_flows callback here as well? That should cover most of the core code.

File: docs/dev/callstacks.qmd [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+# | code-fold: true
+using OrdinaryDiffEqCore: get_du
+model = Ribasim.Model(toml_path)
+du = get_du(model.integrator)
+(; u, p, t) = model.integrator
+graph, verts = tracecall((Ribasim,), Ribasim.water_balance!, (du, u, p, t))
+plot_graph(graph, size = (1700, 1000), xlims = (-0.4, 4.5))
+```
+
+## Allocation initialization
```
Comment (evetion) [inline docs/dev/callstacks.qmd:374]: Can you provide a sentence or two on each figure? Ideally (only) on something that stands out, or is not immediately clear.

**Review by evetion — APPROVED** [review]
Looking good 🚀. Two remarks before we merge:
- Maybe good to make some links from the equations/implementation page to each specific plot?
- In the plots all the Ribasim. prefixes should be removed.

edit:
Maybe remove the spines/axes labels/ticks with [`hidedecorations!`](https://docs.makie.org/dev/reference/blocks/axis#Hiding-Axis-spines-and-decorations) and/or the [minimal theme](https://docs.makie.org/dev/explanations/theming/predefined_themes#theme-minimal).

File: docs/dev/callstacks.qmd (hunk 3 above)
Comment (SouthEndMusic) [reply docs/dev/callstacks.qmd:51]: I decided against adding this because `update_cumulative_flows!` calls very few other (Ribasim) functions.

**Pull request merged**